eth/handler.go

```diff
@@ -632,26 +632,31 @@ func (pm *ProtocolManager) handleMsg(p *peer) error {
 
 		// Mark the peer as owning the block and schedule it for import
 		p.MarkBlock(request.Block.Hash())
-		p.SetHead(request.Block.ParentHash(), request.TD)
 		pm.fetcher.Enqueue(p.id, request.Block)
 
 		// Assuming the block is importable by the peer, but possibly not yet done so,
 		// calculate the head hash and TD that the peer truly must have.
 		var (
 			trueHead = request.Block.ParentHash()
-			trueTD   = request.TD
+			trueTD   = new(big.Int).Sub(request.TD, request.Block.Difficulty())
```

Comment:
I think this changes the semantics of `peer.SetHead`? Why do we need to set `trueTD` to parent TD?

Reply:
Ooops, got it.

Reply:
Hmm I think this line plus L635 is the one that fixed the head sync tracking issue. Can @whilei @tzdybal confirm/explain if I'm wrong? As I see it, other `handler.go` and `blockchain.go` changes are only refactoring and does nothing on the logic (except the `nCompare` issue I mentioned before).

Reply:
I agree with you @sorpaas -- IMO this is the only really relevant change, except the `nCompare` one which I'm about to remove.

Reply:
BTW @sorpaas see remove `nCompare` clause, as well.